### PR TITLE
kola/tests/devcontainer: New test

### PR DESCRIPTION
**Pull request by krnowak**
This is to have a dev container test in the new pipeline. It runs only on qemu.

- [x] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)

### CODE REVIEW

**Comment by krnowak** [comment]
CI: http://jenkins.infra.kinvolk.io:8080/job/container/job/test/2347/

**Comment by krnowak** [comment]
Meh, not enough space on the device, will need to come up with something, maybe mount some local disk?

```
++++ qemu.sh: downloading flatcar_production_image.bin for 3348.0.0+nightly-20220831-2100 (amd64) ++++
+ kola run --board=amd64-usr --parallel=8 --platform=qemu --qemu-bios=/usr/share/qemu/bios-256k.bin --qemu-image=flatcar_production_image.bin --tapfile=/work/__TESTS__/qemu/results-run-1.tap --torcx-manifest=/work/__TESTS__/torcx_manifest.json --qemu-skip-mangle 'devcontainer.*'
2022-09-06T15:39:32Z kola: Using "3348.0.0" as version to filter tests...
=== RUN   devcontainer.docker
=== RUN   devcontainer.systemd-nspawn
--- FAIL: devcontainer.systemd-nspawn (29.42s)
        cluster.go:117: cp: error writing 'flatcar_developer_container.bin': No space left on device
        devcontainer.go:207: main script failed: Process exited with status 1
--- FAIL: devcontainer.docker (29.69s)
        cluster.go:117: cp: error writing 'flatcar_developer_container.bin': No space left on device
        devcontainer.go:207: main script failed: Process exited with status 1
FAIL, output in _kola_temp/qemu-2022-09-06-1539-66
harness: test suite failed
```

**Comment by pothos** [comment]
A test flag in kola to increase the qemu image size would be nice

**Review by pothos — COMMENTED** [review]
We had to disable the mangling in the GitHub Action PR builds because they use lxc containers. I propose we don't increase the size of the backing file but we increase the size of the VM disk directly in `setupDisk` where `qemu-img create` is called and we can pass a new image size by adding a final argument.

**Comment by pothos** [comment]
Adding `size=NEWSIZE` to the options here https://github.com/flatcar-linux/mantle/blob/4371409823df871a691c152eba766b6b981ddf1d/platform/qemu.go#L212 also works and then we would just do it for the disks with a backing file (and only if a new size is passed). For simplicity I would require the final size to be passed by the kola argument instead of trying to fetch the size of the backing file (that would need `qemu-img info --output json` to get `virtual-size`).

**Comment by pothos** [comment]
Splitting it up to avoid losetup is also good ;)

File: platform/machine/qemu/flight.go [hunk 1]
```diff
@@ -82,13 +92,20 @@ func NewFlight(opts *Options) (platform.Flight, error) {
 		if info.Format != "raw" {
```
Comment (pothos) [inline platform/machine/qemu/flight.go:92]: nit: The resize is possible for qcow, just the console modification not

File: platform/qemu.go [hunk 2]
```diff
@@ -76,10 +77,45 @@ func MakeCLDiskTemplate(inputPath string) (output *os.File, result error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("opening %v: %v", outputPath, err)
+	}
+	return output, nil
+}
+
+func resizeRawDisk(disk, extraSize string) error {
+	// Resize the base disk, booting the image should claim the
+	// free space.
+	resize := exec.Command("qemu-img", "resize", "-f", "raw", disk,
```
Comment (pothos) [inline platform/qemu.go:99]:
```suggestion
func resizeDisk(disk, extraSize string) error {
	// Resize the base disk, booting the image should claim the
	// free space.
	resize := exec.Command("qemu-img", "resize", disk,
```
This works for both raw and qcow2

**Comment by krnowak** [comment]
I have for now dropped the changes that added an option for growing the base disk and went with the RAID test way - using the type switch and checking for qemu and qemu-unpriv clusters.

I have saved my previous work and some refactoring in a separate branch, maybe will revisit it at some point.

**Comment by krnowak** [comment]
> Adding `size=NEWSIZE` to the options here
> 
> https://github.com/flatcar-linux/mantle/blob/4371409823df871a691c152eba766b6b981ddf1d/platform/qemu.go#L212
> 
> also works and then we would just do it for the disks with a backing file (and only if a new size is passed). For simplicity I would require the final size to be passed by the kola argument instead of trying to fetch the size of the backing file (that would need `qemu-img info --output json` to get `virtual-size`).

I went with that route, but the size is added to the virtual size, because at the point where I'm doing it, I already have this information. And still no flag, that maybe could be hooked up later.

**Review by pothos — APPROVED** [review]
Thanks

**Comment by krnowak** [comment]
@pothos: Hooking up the kola option was surprisingly easy - https://github.com/flatcar-linux/mantle/commit/d55d115b368132a9675d2643f8c1b20c1ea91170

Want me to add this option to this PR too?

**Comment by pothos** [comment]
Sounds good, one can also use it with kola spawn.

File: kola/tests/devcontainer/devcontainer.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+docker import dev-container-image.tar.gz dev-container:42
+sudo rm -f dev-container-image.tar.gz
+stat /home/core/dev-container-script >&2
+ls -la /home/core >&2
+
+docker run \
+        --env PORTAGE_BINHOST="${PORTAGE_BINHOST}" \
+        --env EXPECTED_VERSION="${EXPECTED_VERSION}" \
+        --mount type=bind,source=/lib/modules,target=/lib/modules,readonly=true \
+        --mount type=bind,source=/home/core/dev-container-script,target=/home/core/dev-container-script,readonly=true \
```
Comment (pothos) [inline kola/tests/devcontainer/devcontainer.go:169]:
```suggestion
        -v /home/core/dev-container-script:/home/core/dev-container-script:z \
```
See https://docs.docker.com/storage/bind-mounts/#configure-the-selinux-label on how to grant access by relabeling

Comment (krnowak) [reply kola/tests/devcontainer/devcontainer.go:174]: It basically modifies the label on the host, so I went with calling restorecon. Let's see if it does the trick.

Comment (krnowak) [reply kola/tests/devcontainer/devcontainer.go:174]: Also, I'm following this: https://github.com/moby/moby/issues/32579#issuecomment-788389016

**Comment by krnowak** [comment]
Tests started to work, but only with the fixed emerge-gitclone (https://github.com/flatcar/flatcar-dev-util/pull/13 and https://github.com/flatcar/coreos-overlay/pull/2166).

**Review by pothos — APPROVED** [review]

**Pull request merged**